### ISSUE TITLE
redundant_clone does not take `as_...()` into account 

### ISSUE THREAD

**Issue by tamaroning** [rust-lang/rust-clippy#8927]
### Summary

redundant_clone does not take as_mut and as_ref into account 

### Reproducer

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0164bff1d54db170c27d2921c8f881d7
I tried this code:

```rust
fn main() {
    let mut a = Box::new(100u32);
    let _ = a.as_mut();
    let _ = a.clone();
}
```

I expected to see this happen:
clippy warns needless clone in line 4

Instead, this happened:
no error

### Additional Labels

@rustbot label +I-false-nagative

**Comment by kyoto7250** [comment]
Hello.

I thought we should not lint.

```rust
fn main() {
    let mut a = Box::new(100u32);
    let r = a.as_mut();
    let x = r.clone();
    println!("{:?}", r);
}
```

The code above is not the same as the code below.

```rust
fn main() {
    let mut a = Box::new(100u32);
    let r = a.as_mut();
    let x = r;
    println!("{:?}", r); # r is moved.
}
```

 As far as I know `let _ = r;` does not move ownership.

```rust
fn main() {
    let mut a = Box::new(100u32);
    let r = a.as_mut();
    let _ = r;
    println!("{:?}", r); # this is ok
}
```

To implement this, we need to check ownership does not move in the end.

**Comment by tamaroning** [comment]
Oh thank you. I have little thought of that.
